app/services/hyrax/indexing/field_mapper.rb

```diff
@@ -0,0 +1,158 @@
+      def set_field(doc, name, value, *indexer_args)
+        # adding defaults indexer
+        indexer_args = [:stored_searchable] if indexer_args.empty?
+        doc.merge! solr_names_and_values(name, value, indexer_args)
+        doc
+      end
+
+      # @api
+      # Given a field name, index_type, etc., returns the corresponding Solr name.
+      # TODO field type is the input format, maybe we could just detect that?
+      # See https://github.com/samvera/active_fedora/issues/1338
+      # @param [String] field_name the ruby (term) name which will get a suffix appended to become a Solr field name
+      # @param opts - index_type is only needed if the FieldDescriptor requires it (e.g. :searcahble)
+      # @return [String] name of the solr field, based on the params
+      def solr_name(field_name, *opts)
+        index_type, args = if opts.first.is_a? Hash
+                             [:stored_searchable, opts.first]
+                           elsif opts.empty?
+                             [:stored_searchable, { type: :text }]
+                           else
+                             [opts[0], opts[1] || { type: :string }]
+                           end
+
+        descriptor = indexer(index_type)
+        names = descriptor.name_and_converter(field_name, args)
+        names.first
+      end
+
+      # Given a field name-value pair, a data type, and an array of index types, returns a hash of
+      # mapped names and values. The values in the hash are _arrays_, and may contain multiple values.
+      def solr_names_and_values(field_name, field_value, index_types)
+        return {} if field_value.nil?
+
+        # Determine the set of index types
+        index_types = Array(index_types)
+        index_types.uniq!
+        index_types.dup.each do |index_type|
+          if index_type.to_s =~ /^not_(.*)/
+            index_types.delete index_type # not_foo
+            index_types.delete Regexp.last_match(1).to_sym # foo
+          end
+        end
+
+        # Map names and values
+
+        results = {}
+
+        # Time seems to extend enumerable, so wrap it so we don't interate over each of its elements.
+        field_value = [field_value] if field_value.is_a? Time
+
+        index_types.each do |index_type|
+          Array(field_value).each do |single_value|
+            # Get mapping for field
+            descriptor = indexer(index_type)
+            data_type = extract_type(single_value)
+            name, _converter = descriptor.name_and_converter(field_name, type: data_type)
+            next unless name
+
+            # Is there a custom converter?
+            # TODO instead of a custom converter, look for input data type and output data type. Create a few methods that can do that cast.
+            # See https://github.com/samvera/active_fedora/issues/1339
+
+            # The converter requires that the Descriptor be constructed with the
+            # :converter option
+            # value = if converter
```

Comment:
I was not able to construct a `Descriptor` with a custom converter `lambda` within this `FieldMapper`, and hence I preserved the note for this issue and kept the block of code.